debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/ehcache/EhcacheLogMinerTransactionCache.java

```diff
@@ -0,0 +1,187 @@
+
+    @Override
+    public EhcacheTransaction getTransaction(String transactionId) {
+        return transactionCache.get(transactionId);
+    }
+
+    @Override
+    public void addTransaction(EhcacheTransaction transaction) {
+        transactionCache.put(transaction.getTransactionId(), transaction);
+        eventIdsByTransactionId.put(transaction.getTransactionId(), new TreeSet<>());
+    }
+
+    @Override
+    public void removeTransaction(EhcacheTransaction transaction) {
+        transactionCache.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean containsTransaction(String transactionId) {
+        return eventIdsByTransactionId.containsKey(transactionId);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return eventIdsByTransactionId.isEmpty();
+    }
+
+    @Override
+    public int getTransactionCount() {
+        return eventIdsByTransactionId.size();
+    }
+
+    @Override
+    public <R> R streamTransactionsAndReturn(Function<Stream<EhcacheTransaction>, R> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            return consumer.apply(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void transactions(Consumer<Stream<EhcacheTransaction>> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void eventKeys(Consumer<Stream<String>> consumer) {
+        try (var stream = StreamSupport.stream(eventCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getKey));
+        }
+    }
+
+    @Override
+    public Iterator<LogMinerEvent> eventsIterator(EhcacheTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId())
+                .stream()
+                .map(eventId -> getTransactionEvent(transaction, eventId))
+                .iterator();
+    }
+
+    @Override
+    public LogMinerEvent getTransactionEvent(EhcacheTransaction transaction, int eventKey) {
+        return eventCache.get(transaction.getEventId(eventKey));
+    }
+
+    @Override
+    public EhcacheTransaction getAndRemoveTransaction(String transactionId) {
+        final EhcacheTransaction transaction = getTransaction(transactionId);
+        if (transaction != null) {
+            transactionCache.remove(transactionId);
+        }
+        return transaction;
+    }
+
+    @Override
+    public void addTransactionEvent(EhcacheTransaction transaction, int eventKey, LogMinerEvent event) {
+        eventCache.put(transaction.getEventId(eventKey), event);
+        eventIdsByTransactionId.get(transaction.getTransactionId()).add(eventKey);
+    }
+
+    @Override
+    public void removeTransactionEvents(EhcacheTransaction transaction) {
+        eventIdsByTransactionId.get(transaction.getTransactionId())
```

Comment:
Wondering if it would be better to removeAll the events from the cache at once.

```
Set<String> eventKeys = eventIdsByTransactionId.get(transaction.getTransactionId())
                .descendingSet()
                .stream()
                .map(transaction::getEventId)
                .collect(Collectors.toSet());
        eventCache.removeAll(eventKeys);
```

Reply:
In this case, we can probably drop the `descendingSet()` too as its superfluous.